docs/src/adr/017adr-temporal.md

```diff
@@ -0,0 +1,190 @@
+
+## Options
+
+SAT encodings of bounded model checking for LTL are a textbook material. A
+linear encoding for LTL is presented in the paper [Biere et al. 2006][]. It is
+explained in [Handbook of Satisfiability][] (Chapter 14). Hence, we do not have
+to do much research around this. However, we have to adapt the standard
+techniques to the domain of TLA+. For instance, we have to understand how to
+deal with `ENABLED`, `WF`, and `SF`, which deviate from the standard setting of
+model checking.
+
+Viktor Sergeev wrote the [first prototype][] for liveness checking at
+University of Lorraine in 2019. Since his implementation was tightly integrated
+with the exploration algorithm, which was refactored several times, this
+implementation has not been integrated in the main branch. We have learned from
+this prototype and discuss our options under this angle.
+
+There are two principally different approaches to the implementation of LTL
+model checking:
+
+ 1. Tight integration with [Transition Executor][].
+
+    In this approach, we would extend the transition executor to incrementally
+    check LTL properties via the encoding by [Biere et al. 2006][]. This approach
+    would let us implement various optimizations. However, it would be harder
+    to maintain and adapt, as we have seen from the first prototype.
+
+ 1. Specification preprocessing.
+
+    In this approach, given a specification `S` and a temporal property `P`, we
+    would produce another specification `S_P` and an invariant `I_P` that has
+    the following property: the temporal property `P` holds on specification
+    `S` if and only if the invariant `I_P` holds on specification `S_P`.
+
+We choose the second approach via specification preprocessing. This will
+simplify maintenance of the codebase. Additionally, the users will be able to
+see the result of preprocessing and optimize it, if needed. When the new
+implementation is well understood, we can revisit it and consider Option 1,
+once [ADR-010][] is implemented.
+
+## Solution
+
+The work plan is tracked in the [issue on temporal properties][].
+
+We propose to split this work into two big subtasks:
+
+ - *Task 1. Temporal operators*:
+    Support for `<>P`, `[]P`, `<A>_e`, and `[A]_e` via preprocessing.
+
+ - *Task 2. Fairness*:
+    Support for `ENABLED A`, `WF_e(A)`, and `SF_e(A)` via preprocessing.
+
+The task on *Temporal operators* is well-understood and poses no technical
+risk. By having solved Task 1, we can give users a relatively complete toolset
+for safety and liveness checking. Indeed, even fairness properties can be
+expressed via `<>` and `[]`.
+
+To support temporal reasoning as it was designed by Leslie Lamport, we have to
+solve Task 2. Most likely, we will have to introduce additional assumptions
+about specifications to solve this task.
+
+### 1. Temporal operators
+
+This task boils down to the implementation of the encoding explained in [Biere
+et al. 2006][]. Explain the lassos.
```

Comment:
> Explain the lassos.

is this supposed to be here?

Reply:
Oops. Good that you have caught it! I wanted to explain lassos a bit.